book/src/e2e-testing.md

```diff
@@ -0,0 +1,81 @@
+# End to End Testing for CSI Storage Plugins
+
+Currently, csi-sanity exists to help test compliance with the spec. But e2e testing of plugins is needed as well to provide plugin authors and users stronger guarantees that their plugin "just works" in kubernetes.
+
+## Setting up End to End tests for your CSI Plugin
+
+### Prerequisites:
+ * A Kubernetes v1.12+ Cluster
+ * [Kubectl](https://kubernetes.io/docs/tasks/tools/install-kubectl/#install-kubectl)
+
+There are two ways to run end-to-end tests for your CSI Plugin
+ 1) use [Kubernetes E2E Tests](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/storage/external), by providing a DriverDefinition YAML file via a parameter.
+ 2) import the in-tree storage tests and run them using `go test`. 
+ 
+ This doc will cover how to run the E2E tests using the second method. 
+ 
+### Note: 
+ 
+ In some cases you would not be able to use the first method, in running e2e tests by just providing a YAML file defining your CSI plugin.
+
+ * For example the [NFS CSI plugin](https://github.com/kubernetes-csi/csi-driver-nfs) does not support dynamic provisoning, so we would want to skip those and run only pre-provisioned tests. For such cases, you would need to write your own testdriver, which is discussed below. 
+
+In-tree storage e2e tests could be used to test CSI storage plugins. Your repo could be setup similar to how this [NFS CSI plugin](https://github.com/mathu97/csi-driver-nfs/) is setup, where the testfiles are in a test directory and the main test file is in the cmd directory. 
+
+To be able import the in-tree storage tests, CSI plugin authors would need to implement a [testdriver](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L31) for their CSI plugin. The testdriver provides required functionality that would help setup testcases for a particular plugin. 
+
+For any testdriver the basic required functions are (Since it implements the [TestDriver Interface](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/testsuites/testdriver.go#L31)):
+ - `GetDriverInfo() *testsuites.DriverInfo`
+ - `SkipUnsupportedTest(pattern testpatterns.TestPattern)`
+ - `PrepareTest(f *framework.Framework) (*testsuites.PerTestConfig, func())` 
+ 
+The `PrepareTest` method is important as it is where you will write code to deploy your driver and setup other pre-requistes that your driver requires. The nfs and hostpath testdrivers deploy the plugin by simply [using the manifests files](https://github.com/mathu97/csi-driver-nfs/blob/master/test/nfs_driver.go#L91) that you would regularly use to deploy a plugin. 
```

Comment:
In tree we bring up/down the driver per test case.  For running out-of-tree in a driver-specific repo, I would encourage folks to bring up their driver at the beginning/before the test instead of per-test.